test-data/unit/check-literal-string.test

```diff
@@ -0,0 +1,250 @@
+from typing_extensions import LiteralString
+
+class Base:
+    x: int
+class Sub(Base):
+    x = 'a'  # E: Incompatible types in assignment (expression has type "str", base class "Base" defined the type as "int")
+class ExcplicitSub1(Base):
+    x: LiteralString = 'a'  # E: Incompatible types in assignment (expression has type "LiteralString", base class "Base" defined the type as "int")
+class ExcplicitSub2(Base):
+    x: LiteralString  # E: Incompatible types in assignment (expression has type "LiteralString", base class "Base" defined the type as "int")
+
+def accepts_int(arg: int): ...
+accepts_int('b')  # E: Argument 1 to "accepts_int" has incompatible type "str"; expected "int"
+ls1: LiteralString
+ls2: LiteralString = 'a'
+accepts_int(ls1)  # E: Argument 1 to "accepts_int" has incompatible type "LiteralString"; expected "int"
+accepts_int(ls2)  # E: Argument 1 to "accepts_int" has incompatible type "LiteralString"; expected "int"
+[builtins fixtures/literal_string.pyi]
+
+
+[case testConditionalLiteralString]
+from typing_extensions import LiteralString
+if int():
+    cond: LiteralString = 'abc'
+else:
+    cond = ','.join(['a', 'b'])  # E: Incompatible types in assignment (expression has type "str", variable has type "LiteralString")
+[builtins fixtures/literal_string.pyi]
+
+
+[case testUnionOfLiteralString]
+from typing_extensions import LiteralString
+from typing import Union
+
+x: Union[str, LiteralString] = 'a'
+y: Union[str, LiteralString] = str()
+z: Union[str, LiteralString] = 1  # E: Incompatible types in assignment (expression has type "int", variable has type "Union[str, LiteralString]")
+[builtins fixtures/literal_string.pyi]
+
+
+[case testTypeAliasOfLiteralString]
+from typing_extensions import LiteralString, TypeAlias
+
+LS1: TypeAlias = LiteralString
+LS2 = LiteralString
+
+def ls1(arg: LS1) -> LS1: ...
+def ls2(arg: LS2) -> LS2: ...
+
+reveal_type(ls1('abc'))  # N: Revealed type is "LiteralString"
+reveal_type(ls2("abc"))  # N: Revealed type is "LiteralString"
+
+x: LiteralString
+reveal_type(ls1(x))  # N: Revealed type is "LiteralString"
+reveal_type(ls2(x))  # N: Revealed type is "LiteralString"
+y: str
+ls1(y)  # E: Argument 1 to "ls1" has incompatible type "str"; expected "LiteralString"
+ls2(y)  # E: Argument 1 to "ls2" has incompatible type "str"; expected "LiteralString"
+[builtins fixtures/literal_string.pyi]
+
+
+[case testOverloadsWithLiteralString]
+from over import some
+
+x: str
+y: str
+reveal_type(some('a', 'b'))  # N: Revealed type is "LiteralString"
+reveal_type(some(x, y))      # N: Revealed type is "builtins.str"
+reveal_type(some('a', y))    # N: Revealed type is "builtins.str"
+reveal_type(some(x, 'b'))    # N: Revealed type is "builtins.str"
+
+[file over.pyi]
+from typing_extensions import LiteralString
+from typing import Union, overload
+
+@overload
+def some(x: LiteralString, y: LiteralString) -> LiteralString: ...
+@overload
+def some(x: str, y: str) -> str: ...
```

Comment:
Test also a `str` method with `self` annotated as `self: LiteralString` in one overload variant (e.g. for `__add__`)? This way we should be able to make `'x' + 'y'` be a LiteralString without special casing?

Reply:
I actually tried this and mypy gives an error like `Self type is not a supertype of class`. I think this is too strict, we already check self-type at call site. I would say we can simply skip this check for overloads (maybe only keep it for overload implementation).

Reply:
Just to be clear, I think we still need this test, see also my comment above https://github.com/python/mypy/pull/13664#discussion_r970642740